models/user_heatmap.go

```diff
@@ -0,0 +1,40 @@
+// Copyright 2018 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.package models
+
+package models
+
+import (
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/util"
+)
+
+// UserHeatmapData represents the data needed to create a heatmap
+type UserHeatmapData struct {
+	Timestamp     util.TimeStamp `json:"timestamp"`
+	Contributions int64          `json:"contributions"`
+}
+
+// GetUserHeatmapDataByUser returns an array of UserHeatmapData
+func GetUserHeatmapDataByUser(user *User) (hdata []*UserHeatmapData, err error) {
+	var groupBy string
+	switch {
+	case setting.UseSQLite3:
+		groupBy = "CAST(strftime('%Y-%m-%d', created_unix, 'unixepoch') AS INT)"
+	case setting.UseMySQL:
+		groupBy = "UNIX_TIMESTAMP(DATE_FORMAT(FROM_UNIXTIME(created_unix), '%Y%m%d'))"
+	case setting.UsePostgreSQL:
+		groupBy = "EXTRACT(epoch from TO_TIMESTAMP(TO_CHAR(TO_TIMESTAMP(created_unix), 'DD Mon YYYY'), 'DD Mon YYYY'))"
```

Comment:
I think this would be more optimal for Postgres:
```
extract(epoch from date_trunc('day', to_timestamp(created_unix)))
```

Reply:
Fixed.